fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java

```diff
@@ -306,14 +206,119 @@ private void performScan() throws IOException {
+                            limitRemaining--;
+                        }
+                        return RecordCursorResult.withNextValue(result.get().indexEntry, continuationFromDoc);
+                    } else if (exhausted) {
+                        return RecordCursorResult.exhausted();
+                    } else if (limitRemaining <= 0) {
+                        RecordCursorContinuation continuationFromDoc = LuceneCursorContinuation.fromScoreDoc(searchAfter);
+                        return RecordCursorResult.withoutNextValue(continuationFromDoc, NoNextReason.RETURN_LIMIT_REACHED);
+                    } else {
+                        final Optional<NoNextReason> stoppedReason = limitManager.getStoppedReason();
+                        if (!stoppedReason.isPresent()) {
+                            throw new RecordCoreException("limit manager stopped LuceneRecordCursor but did not report a reason");
+                        } else {
+                            return RecordCursorResult.withoutNextValue(LuceneCursorContinuation.fromScoreDoc(searchAfter), stoppedReason.get());
+                        }
+                    }
+                });
         if (timer != null) {
             timer.recordSinceNanoTime(LuceneEvents.Events.LUCENE_INDEX_SCAN, startTime);
-            timer.increment(LuceneEvents.Counts.LUCENE_SCAN_MATCHED_DOCUMENTS, topDocs.scoreDocs.length);
+            timer.increment(LuceneEvents.Counts.LUCENE_SCAN_MATCHED_DOCUMENTS, newTopDocs.scoreDocs.length);
         }
     }
 
+    private CompletableFuture<ScoreDocAndIndexEntry> buildIndexEntryFromScoreDocAsync(@Nonnull ScoreDoc scoreDoc) {
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                Document document = searcher.doc(scoreDoc.doc);
+                IndexableField primaryKey = document.getField(LuceneIndexMaintainer.PRIMARY_KEY_FIELD_NAME);
+                BytesRef pk = primaryKey.binaryValue();
+                if (LOGGER.isTraceEnabled()) {
+                    LOGGER.trace("document={}", document);
+                    LOGGER.trace("primary key read={}", Tuple.fromBytes(pk.bytes, pk.offset, pk.length));
+                }
+                if (timer != null) {
+                    timer.increment(FDBStoreTimer.Counts.LOAD_SCAN_ENTRY);
+                }
+                Tuple setPrimaryKey = Tuple.fromBytes(pk.bytes);
+                // Initialized with values that aren't really legal in a Tuple to find offset bugs.
+                List<Object> fieldValues = Lists.newArrayList(fields);
+                if (groupingKey != null) {
+                    // Grouping keys are at the front.
+                    for (int i = 0; i < groupingKey.size(); i++) {
+                        fieldValues.set(i, groupingKey.get(i));
+                    }
+                }
+                if (storedFields != null) {
+                    for (int i = 0; i < storedFields.size(); i++) {
```

Comment:
I think this is copied over from the previous code, and it wasn't changed, but this looks wrong to me. In particular, the position `i` indicates the index of the _stored_ field array, but then the value is placed into the `fieldValues` array, and I don't think those need to align.

I don't think we necessarily need to block merging this PR on that, though, especially if this bug is preëxisting.

Reply:
`storedFields` is sparse, I believe.

Reply:
I see. If it aligns, there may not be a problem, then